File: app/src/main/java/dev/ragnarok/fenrir/model/Keyboard.java
package dev.ragnarok.fenrir.model;

import android.os.Parcel;
import android.os.Parcelable;

import java.util.ArrayList;
import java.util.List;

public class Keyboard implements Parcelable {

    public static final Creator<Keyboard> CREATOR = new Creator<Keyboard>() {
        @Override
        public Keyboard createFromParcel(Parcel in) {
            return new Keyboard(in);
        }

        @Override
        public Keyboard[] newArray(int size) {
            return new Keyboard[size];
        }
    };
    private boolean one_time;
    private boolean inline;
    private int author_id;
    private List<List<Button>> buttons;

    public Keyboard() {
    }

    protected Keyboard(Parcel in) {
        one_time = in.readByte() != 0;
        inline = in.readByte() != 0;
        author_id = in.readInt();

        int size = in.readInt();
        buttons = new ArrayList<>(size);
        for (int i = 0; i < size; i++) {
            buttons.add(i, in.createTypedArrayList(Button.CREATOR));
        }
    }

    public boolean getOne_time() {
        return one_time;
    }

    public Keyboard setOne_time(boolean one_time) {
        this.one_time = one_time;
        return this;
    }

    public boolean getInline() {
        return inline;
    }

    public Keyboard setInline(boolean inline) {
        this.inline = inline;
        return this;
    }

    public int getAuthor_id() {
        return author_id;
    }

    public Keyboard setAuthor_id(int author_id) {
        this.author_id = author_id;
        return this;
    }

    public List<List<Button>> getButtons() {
        return buttons;
    }

    public Keyboard setButtons(List<List<Button>> buttons) {
        this.buttons = buttons;
        return this;
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeByte((byte) (one_time ? 1 : 0));
        dest.writeByte((byte) (inline ? 1 : 0));
        dest.writeInt(author_id);

        dest.writeInt(buttons.size());
        for (List<Button> i : buttons) {
            dest.writeTypedList(i);
        }
    }

    public static class Button implements Parcelable {

        public static final Creator<Button> CREATOR = new Creator<Button>() {
            @Override
            public Button createFromParcel(Parcel in) {
                return new Button(in);
            }

            @Override
            public Button[] newArray(int size) {
                return new Button[size];
            }
        };
        private String color;
        private String type;
        private String label;
        private String link;
        private String payload;

        public Button() {
        }

        protected Button(Parcel in) {
            color = in.readString();
            type = in.readString();
            label = in.readString();
            link = in.readString();
            payload = in.readString();
        }

        public String getColor() {
            return color;
        }

        public Button setColor(String color) {
            this.color = color;
            return this;
        }

        public String getType() {
            return type;
        }

        public Button setType(String type) {
            this.type = type;
            return this;
        }

        public String getLabel() {
            return label;
        }

        public Button setLabel(String label) {
            this.label = label;
            return this;
        }

        public String getLink() {
            return link;
        }

        public Button setLink(String link) {
            this.link = link;
            return this;
        }

        public String getPayload() {
            return payload;
        }

        public Button setPayload(String payload) {
            this.payload = payload;
            return this;
        }

        @Override
        public int describeContents() {
            return 0;
        }

        @Override
        public void writeToParcel(Parcel dest, int flags) {
            dest.writeString(color);
            dest.writeString(type);
            dest.writeString(label);
            dest.writeString(link);
            dest.writeString(payload);
        }
    }
}

File: app_fenrir/src/main/kotlin/dev/ragnarok/fenrir/model/Keyboard.kt
package dev.ragnarok.fenrir.model

import android.os.Parcel
import android.os.Parcelable
import dev.ragnarok.fenrir.getBoolean
import dev.ragnarok.fenrir.orZero
import dev.ragnarok.fenrir.putBoolean

class Keyboard : Parcelable {
    var one_time = false
        private set
    var inline = false
        private set
    var author_id = 0
        private set
    var buttons: ArrayList<List<Button>>? = null
        private set

    constructor()
    internal constructor(parcel: Parcel) {
        one_time = parcel.getBoolean()
        inline = parcel.getBoolean()
        author_id = parcel.readInt()
        val size = parcel.readInt()
        buttons = ArrayList(size)
        for (i in 0 until size) {
            parcel.createTypedArrayList(Button.CREATOR)?.let { buttons?.add(i, it) }
        }
    }

    fun setOne_time(one_time: Boolean): Keyboard {
        this.one_time = one_time
        return this
    }

    fun setInline(inline: Boolean): Keyboard {
        this.inline = inline
        return this
    }

    fun setAuthor_id(author_id: Int): Keyboard {
        this.author_id = author_id
        return this
    }

    fun setButtons(buttons: ArrayList<List<Button>>?): Keyboard {
        this.buttons = buttons
        return this
    }

    override fun describeContents(): Int {
        return 0
    }

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.putBoolean(one_time)
        dest.putBoolean(inline)
        dest.writeInt(author_id)
        if (buttons == null) {
            dest.writeInt(0)
        } else {
            dest.writeInt(buttons?.size.orZero())
            for (i in buttons.orEmpty()) {
                dest.writeTypedList(i)
            }
        }
    }

    class Button : Parcelable {
        var color: String? = null
            private set
        var type: String? = null
            private set
        var label: String? = null
            private set
        var link: String? = null
            private set
        var payload: String? = null
            private set

        constructor()
        internal constructor(parcel: Parcel) {
            color = parcel.readString()
            type = parcel.readString()
            label = parcel.readString()
            link = parcel.readString()
            payload = parcel.readString()
        }

        fun setColor(color: String?): Button {
            this.color = color
            return this
        }

        fun setType(type: String?): Button {
            this.type = type
            return this
        }

        fun setLabel(label: String?): Button {
            this.label = label
            return this
        }

        fun setLink(link: String?): Button {
            this.link = link
            return this
        }

        fun setPayload(payload: String?): Button {
            this.payload = payload
            return this
        }

        override fun describeContents(): Int {
            return 0
        }

        override fun writeToParcel(dest: Parcel, flags: Int) {
            dest.writeString(color)
            dest.writeString(type)
            dest.writeString(label)
            dest.writeString(link)
            dest.writeString(payload)
        }

        companion object CREATOR : Parcelable.Creator<Button> {
            override fun createFromParcel(parcel: Parcel): Button {
                return Button(parcel)
            }

            override fun newArray(size: Int): Array<Button?> {
                return arrayOfNulls(size)
            }
        }
    }

    companion object CREATOR : Parcelable.Creator<Keyboard> {
        override fun createFromParcel(parcel: Parcel): Keyboard {
            return Keyboard(parcel)
        }

        override fun newArray(size: Int): Array<Keyboard?> {
            return arrayOfNulls(size)
        }
    }
}